Provide an option in the test verb to return an error code on test failure

From discussion https://github.com/colcon/colcon-cmake/pull/3#issuecomment-382590816

I tried removing a header in rcutils and got a non-zero return code.

```bash
--- stderr: rcutils
/tmp/test_colcon/src/ros2/rcutils/src/filesystem.c:19:32: fatal error: rcutils/filesystem.h: No such file or directory
compilation terminated.
make[2]: *** [CMakeFiles/rcutils.dir/src/filesystem.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/rcutils.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
---
Failed   <<< rcutils	[ Exited with code 2 ]

Summary: 0 packages finished [7.01s]
  1 package failed: rcutils
  1 package had stderr output: rcutils

$ echo $?
2
```

Can you please open a separate issue with a ready to run, reproducible example?

Hmm, it could probably be some wrong config on my end? I'll create a ticket
if I can't fix it. Thanks.

On Wed, Sep 19, 2018 at 10:28 AM Mikael Arguedas <notifications@github.com>
wrote:

> I tried removing a header in rcutils and got a non-zero return code.
>
> --- stderr: rcutils
> /tmp/test_colcon/src/ros2/rcutils/src/filesystem.c:19:32: fatal error: rcutils/filesystem.h: No such file or directory
> compilation terminated.
> make[2]: *** [CMakeFiles/rcutils.dir/src/filesystem.c.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> make[1]: *** [CMakeFiles/rcutils.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make: *** [all] Error 2
> ---
> Failed   <<< rcutils	[ Exited with code 2 ]
>
> Summary: 0 packages finished [7.01s]
>   1 package failed: rcutils
>   1 package had stderr output: rcutils
>
> $ echo $?
> 2
>
> Can you please open a separate issue with a ready to run, reproducible
> example?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/colcon/colcon-core/issues/21#issuecomment-422889797>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEl9x8rgmlGksQtTOD9LoKTiekSepo7cks5ucn67gaJpZM4TcQky>
> .
>

--- a/colcon_core/executor/__init__.py
+++ b/colcon_core/executor/__init__.py
@@ -60,7 +60,7 @@ class Job:
         """
         self._event_queue = event_queue
 
-        self._put_event_into_queue(JobQueued(
+        self.put_event_into_queue(JobQueued(
             self.task_context.pkg.name, self.task_context.dependencies))
 
     async def __call__(self, *args, **kwargs):
@@ -80,10 +80,10 @@ class Job:
         :returns: The return code of the invoked task
         :raises Exception: Any exception the invoked task raises
         """
-        self._put_event_into_queue(JobStarted(self.task_context.pkg.name))
+        self.put_event_into_queue(JobStarted(self.task_context.pkg.name))
 
         # replace function to use this job as the event context
-        self.task_context.put_event_into_queue = self._put_event_into_queue
+        self.task_context.put_event_into_queue = self.put_event_into_queue
         self.task.set_context(context=self.task_context)
 
         rc = 0
@@ -93,17 +93,22 @@ class Job:
             rc = SIGINT_RESULT
         except Exception:
             rc = 1
-            self._put_event_into_queue(
+            self.put_event_into_queue(
                 StderrLine(traceback.format_exc().encode()))
             raise
         finally:
             if self.returncode is None:
                 self.returncode = rc or 0
-            self._put_event_into_queue(
+            self.put_event_into_queue(
                 JobEnded(self.task_context.pkg.name, self.returncode))
         return self.returncode
 
-    def _put_event_into_queue(self, event):
+    def put_event_into_queue(self, event):
+        """
+        Post a message event into the event queue.
+
+        :param event: The event
+        """
         self._event_queue.put((event, self))
 
     def __str__(self):

--- a/colcon_core/verb/test.py
+++ b/colcon_core/verb/test.py
@@ -5,9 +5,11 @@ from collections import OrderedDict
 import copy
 import logging
 import os
+import types
 
 from colcon_core.argument_parser.destination_collector \
     import DestinationCollectorDecorator
+from colcon_core.event.test import TestFailure
 from colcon_core.event_handler import add_event_handler_arguments
 from colcon_core.executor import add_executor_arguments
 from colcon_core.executor import execute_jobs
@@ -168,6 +170,10 @@ class TestVerb(VerbExtensionPoint):
             action='store_true',
             help='Abort after the first package with any errors (failing '
                  'tests are not considered errors in this context)')
+        parser.add_argument(
+            '--return-code-on-test-failure',
+            action='store_true',
+            help='Use a non-zero return code to indicate any test failure')
         add_executor_arguments(parser)
         add_event_handler_arguments(parser)
 
@@ -196,9 +202,35 @@ class TestVerb(VerbExtensionPoint):
             os.getcwd(), context.args.install_base))
         jobs = self._get_jobs(context.args, decorators, install_base)
 
+        if context.args.return_code_on_test_failure:
+            # watch published events on all jobs to detect any test failures
+            any_test_failures = False
+
+            def check_for_test_failures(put_event_into_queue):
+                nonlocal any_test_failures
+
+                def put_event_into_queue_(self, event):
+                    nonlocal any_test_failures
+                    nonlocal put_event_into_queue
+                    if isinstance(event, TestFailure):
+                        any_test_failures = True
+                    return put_event_into_queue(event)
+
+                return put_event_into_queue_
+
+            for job in jobs.values():
+                job.put_event_into_queue = types.MethodType(
+                    check_for_test_failures(job.put_event_into_queue), job)
+
         on_error = OnError.continue_ \
             if not context.args.abort_on_error else OnError.interrupt
-        return execute_jobs(context, jobs, on_error=on_error)
+        rc = execute_jobs(context, jobs, on_error=on_error)
+
+        if context.args.return_code_on_test_failure:
+            if not rc and any_test_failures:
+                return 1
+
+        return rc
 
     def _get_jobs(self, args, decorators, install_base):
         jobs = OrderedDict()

